boards/arm/frdm_k64f/pinmux.c

```diff
@@ -155,6 +155,13 @@ static int frdm_k64f_pinmux_init(struct device *dev)
 		       PORT_PCR_PE_MASK | PORT_PCR_PS_MASK);
 #endif
 
+#if CONFIG_SHIELD_ADAFRUIT_WINC1500
+	/* IRQ, ENable, RST */
+	pinmux_pin_set(portc,  3, PORT_PCR_MUX(kPORT_MuxAsGpio));
+	pinmux_pin_set(portc,  2, PORT_PCR_MUX(kPORT_MuxAsGpio));
+	pinmux_pin_set(porta,  2, PORT_PCR_MUX(kPORT_MuxAsGpio));
+#endif
```

Comment:
If shield is truly arduino compatible, this shouldn't be needed. If any modification required, these pins should be enabled as part of default board configuration.
Cf https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#general-recommendations:
"Enable all GPIO ports connected to peripherals or expansion connectors."

Reply:
I wont say I disagree, but this pinmux.c has a several other GPIO's being configured based on other configs:
MODEM_WNCM14A2A, MODEM_UBLOX_SARA_R4.

I know this doesn't necessarily make my change valid, but currently the default board configuration doesn't (clearly / to my knowledge) enable those pins. 
Isn't such cleanup of the board configuration another PR?

Reply:
Is it something like this you think is the intention with the board_porting guideline

`diff --git a/boards/arm/frdm_k64f/pinmux.c b/boards/arm/frdm_k64f/pinmux.c
index 8e96be3704..61f203af4f 100644
--- a/boards/arm/frdm_k64f/pinmux.c
+++ b/boards/arm/frdm_k64f/pinmux.c
@@ -33,6 +33,31 @@ static int frdm_k64f_pinmux_init(struct device *dev)
                device_get_binding(CONFIG_PINMUX_MCUX_PORTE_NAME);
 #endif
 
+       /* All exposed Arduino compatible pins defaults to GPIO */
+       pinmux_pin_set(portc, 16, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D0 */
+       pinmux_pin_set(portc, 17, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D1 */
+       pinmux_pin_set(portb,  9, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D2 */
+       pinmux_pin_set(porta,  1, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D3 */
+       pinmux_pin_set(portb, 23, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D4 */
+       pinmux_pin_set(porta,  2, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D5 */
+       pinmux_pin_set(portc,  2, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D6 */
+       pinmux_pin_set(portc,  3, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D7 */
+       pinmux_pin_set(portc, 12, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D8 */
+       pinmux_pin_set(portc,  4, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D9 */
+       pinmux_pin_set(portd,  0, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D10 */
+       pinmux_pin_set(portd,  2, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D11 */
+       pinmux_pin_set(portd,  3, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D12 */
+       pinmux_pin_set(portd,  1, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D13 */
+       pinmux_pin_set(porte, 25, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D14 */
+       pinmux_pin_set(porte, 25, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* D15 */
+
+       pinmux_pin_set(portb,  2, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* A0 */
+       pinmux_pin_set(portb,  3, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* A1 */
+       pinmux_pin_set(portb, 10, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* A2 */
+       pinmux_pin_set(portb, 11, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* A3 */
+       pinmux_pin_set(portc, 11, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* A4 */
+       pinmux_pin_set(portc, 10, PORT_PCR_MUX(kPORT_MuxAsGpio)); /* A5 */
+
 #if DT_NODE_HAS_STATUS(DT_NODELABEL(uart0), okay) && CONFIG_SERIAL
        /* UART0 RX, TX */
        pinmux_pin_set(portb, 16, PORT_PCR_MUX(kPORT_MuxAlt3));
@@ -66,41 +91,6 @@ static int frdm_k64f_pinmux_init(struct device *dev)
        pinmux_pin_set(porte, 26, PORT_PCR_MUX(kPORT_MuxAsGpio));
        pinmux_pin_set(portb, 21, PORT_PCR_MUX(kPORT_MuxAsGpio));
 
-#ifdef CONFIG_MODEM_WNCM14A2A
-       /* WNC-M14A2A Modem POWER_ON */
-       pinmux_pin_set(portb,  9, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* WNC-M14A2A Shield PMOD_D1 */
-       pinmux_pin_set(portb, 10, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* WNC-M14A2A Shield PMOD_D2 */
-       /* WNC-M14A2A Shield PMOD_D2 */
-       pinmux_pin_set(portb, 11, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* WNC-M14A2A Modem WWAN_STATE */
-       pinmux_pin_set(portb, 23, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* WNC-M14A2A Modem WAKEUP_ENABLE */
-       pinmux_pin_set(portc,  2, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* WNC-M14A2A Modem HTS221_DRDY */
-       pinmux_pin_set(portc,  3, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* WNC-M14A2A Shield LEVEL_TRANSFORM_ENABLE */
-       pinmux_pin_set(portc,  4, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* WNC-M14A2A Modem RESET */
-       pinmux_pin_set(portc, 12, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* WNC-M14A2A Modem BOOT_MODE_SELECT */
-       pinmux_pin_set(portc, 17, PORT_PCR_MUX(kPORT_MuxAsGpio));
-#endif
-
-#ifdef CONFIG_MODEM_UBLOX_SARA_R4
-       /* Modem RESET */
-       pinmux_pin_set(portc,  2, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* Modem POWER_ON */
-       pinmux_pin_set(porta,  2, PORT_PCR_MUX(kPORT_MuxAsGpio));
-#endif
-
-#ifdef CONFIG_IEEE802154_MCR20A
-       /* FRDM-MCR20A Reset (D5) */
-       pinmux_pin_set(porta,  2, PORT_PCR_MUX(kPORT_MuxAsGpio));
-       /* FRDM-MCR20A IRQ_B (D2) */
-       pinmux_pin_set(portb,  9, PORT_PCR_MUX(kPORT_MuxAsGpio));
-#endif
-
 #if DT_NODE_HAS_STATUS(DT_NODELABEL(spi0), okay) && CONFIG_SPI
        /* SPI0 CS0, SCK, SOUT, SIN */
`

Reply:
Ok, I hadn't seen the issue was that deep.
So, yes I think this is indeed be the way things should be done. Now considering the potential conflicts this would raise, going this direction is like opening a can of worms.

So IIUC, to use a GPIO through normal GPIO API, pin should first be set as GPIO in this file. I wonder why this is not done directly in gpio driver. Pin conflict handling maybe ? @MaureenHelm maybe you could shed some light here ?

Reply:
Yes, pin conflicts are the issue. We need to get to the point where we can configure pinmuxes in device tree, but I don't want to solve that problem in this PR. I think what's currently implemented in the PR is ok for now.